File: server/app/services/freshdesk/canned_response_service.rb
module Freshdesk
    class CannedResponseService
        def initialize
            @client = ApiClient.new
        end

        def list_folders
            begin
                @client.request(:get, '/canned_response_folders')
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to list canned response folders: #{e.message}")
            end
        end

        def get_folder(id)
            begin
                @client.request(:get, "/canned_response_folders/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Canned response folder ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get canned response folder ##{id}: #{e.message}")
            end
        end

        def get_response(id)
            begin
                @client.request(:get, "/canned_responses/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Canned response ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get canned response ##{id}: #{e.message}")
            end
        end
    end
end

File: client/src/services/cannedResponseService.js
import axios from 'axios';
import { handleApiError } from '../utils/errorHandler';

const API_URL = import.meta.env.VITE_API_URL;

const cannedResponseService = {
    getFolders: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/canned_response_folders`, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching canned response folders:", error);
            throw handleApiError(error);
        }
    },

    getFolderResponses: async (folderId) => {
        try {
            const response = await axios.get(`${API_URL}/api/canned_response_folders/${folderId}`, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error(`Error fetching responses for folder ${folderId}:`, error);
            throw handleApiError(error);
        }
    },

    getResponse: async (responseId) => {
        try {
            const response = await axios.get(`${API_URL}/api/canned_responses/${responseId}`, {
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error(`Error fetching canned response ${responseId}:`, error);
            throw handleApiError(error);
        }
    }
};

export default cannedResponseService;
